src/coap_blockwise.c

```diff
@@ -63,65 +69,113 @@ static void on_block_sent(struct golioth_client *client,
+        // Make sure the block size is a valid size for blockwise uploads
+        if (!IS_VALID_BLOCK_SIZE(block_size))
+        {
+            return GOLIOTH_ERR_INVALID_BLOCK_SIZE;
+        }
+        ctx->block_size = block_size;
+    }
+    else if (block_size != ctx->block_size)
+    {
+        // Size of all blocks must be the same
+        return GOLIOTH_ERR_INVALID_BLOCK_SIZE;
+    }
+
+    return GOLIOTH_OK;
 }
 
 // Function to upload a single block
-static enum golioth_status upload_single_block(struct golioth_client *client,
-                                               struct blockwise_transfer *ctx)
+static void upload_single_block(struct golioth_client *client, struct blockwise_transfer *ctx)
 {
-    assert(ctx->block_size > 0 && ctx->block_size <= CONFIG_GOLIOTH_BLOCKWISE_UPLOAD_BLOCK_SIZE);
-    return golioth_coap_client_set_block(client,
-                                         ctx->path_prefix,
-                                         ctx->path,
-                                         ctx->is_last,
-                                         GOLIOTH_CONTENT_TYPE_JSON,
-                                         ctx->offset,
-                                         ctx->block_buffer,
-                                         ctx->block_size,
-                                         on_block_sent,
-                                         ctx,
-                                         true,
-                                         GOLIOTH_SYS_WAIT_FOREVER);
+    enum golioth_status status = golioth_coap_client_set_block(client,
+                                                               ctx->path_prefix,
+                                                               ctx->path,
+                                                               ctx->is_last,
+                                                               ctx->content_type,
+                                                               ctx->offset,
+                                                               ctx->block_buffer,
+                                                               ctx->block_size,
+                                                               on_block_sent,
+                                                               ctx,
+                                                               true,
+                                                               GOLIOTH_SYS_WAIT_FOREVER);
+    ctx->status = status;
 }
 
 // Function to handle blockwise upload errors
-static enum golioth_status handle_upload_error()
+// This function retries for all errors like invalid state, failed mem alloc,
+// queue full, disconnects, etc returned by golioth_coap_client_set_block
+static enum golioth_status handle_upload_error(struct golioth_client *client,
+                                               struct blockwise_transfer *ctx)
 {
-    // TODO: handle errors like disconnects etc
-    return GOLIOTH_OK;
+    while (1)
```

Comment:
We should check (and decrement) the retry count here in the `while()`. I also don't think we need to keep `retry_count` in the context struct.